File: apps/desktop/src/lib/crash.ts
// apps/desktop/src/lib/crash.ts
//
// Crash reporting tier for the desktop renderer (P3 of
// docs/observability-analytics-support-access.md). This is a SEPARATE consent tier
// from product analytics (analytics.ts) — the Zed/VS Code/Cursor split — gated on
// the canonical `crash-reports-enabled` pref, not `product-analytics-enabled`.
//
// What it captures: unhandled renderer errors (React error boundary +
// window.onerror / unhandledrejection). NOT product events, NOT prompts, NOT agent
// content. The Rust panic tier (Core + Gateway) is the sibling; see
// apps/core/src/crash.rs.
//
// Privacy posture (the load-bearing AC — the leak vector is Sentry's DEFAULT
// integrations, not the top-level event fields):
//   - sendDefaultPii: false, and beforeSend drops event.request / event.user /
//     event.server_name (Sentry captures the machine hostname by default).
//   - beforeBreadcrumb drops console / fetch / xhr breadcrumbs — console.* and
//     fetch URLs can carry content. No Session Replay, no profiling, no tracing.
//   - A home-dir-ish scrub on the message + frame paths (defense in depth; the
//     renderer rarely has absolute home paths, but scrub anyway).
//   - Anonymous: the random, account-unlinked install id from analytics.ts is the
//     only id; setUser is never called.
//
// Opt-out gate (matches analytics.ts exactly):
//   - emit/report is a no-op unless BOTH a DSN is configured AND the in-memory
//     enabled flag is true. The flag is flipped live by setCrashReportingEnabled()
//     from the Privacy toggle, which beforeSend honors immediately (returns null
//     when off) so toggling off stops egress without an async Core re-read.
//   - Seeded synchronously from a localStorage mirror of `crash-reports-enabled`.
//
// Nothing hardcoded: the DSN comes from VITE_SENTRY_DSN. With no DSN the whole
// module is a graceful no-op (no vendor wired).

import {
	captureException,
	type ErrorEvent as SentryErrorEvent,
	init as sentryInit,
} from "@sentry/react";
import { getInstallId } from "./analytics.ts";

// --- Config (swappable, env-driven) -----------------------------------------

const SENTRY_DSN = import.meta.env.VITE_SENTRY_DSN as string | undefined;

// localStorage key mirroring the canonical kebab pref name, under the `ryu:`
// namespace — synchronous seed before the async Core read resolves.
const ENABLED_MIRROR_KEY = "ryu:crash-reports-enabled";

// --- Runtime state + gate ---------------------------------------------------

let initialized = false;
let enabled = readEnabledMirror();

function readEnabledMirror(): boolean {
	try {
		// Default ON (opt-out posture). Only an explicit "false" opts out.
		return localStorage.getItem(ENABLED_MIRROR_KEY) !== "false";
	} catch {
		return true;
	}
}

function writeEnabledMirror(next: boolean): void {
	try {
		localStorage.setItem(ENABLED_MIRROR_KEY, String(next));
	} catch {
		// no-op
	}
}

/** Whether a Sentry DSN is configured at all. */
export function isCrashReportingConfigured(): boolean {
	return Boolean(SENTRY_DSN);
}

/** The current effective gate. */
export function isCrashReportingEnabled(): boolean {
	return enabled;
}

// --- PII scrub --------------------------------------------------------------

// Match an absolute home-style path prefix so it can be replaced with `~`. Covers
// POSIX (/home/<user>, /Users/<user>) and Windows (C:\Users\<user>). Top-level
// regex (not built in a loop) per the repo perf rule.
const HOME_PATH_RE = /(?:[A-Za-z]:\\Users\\[^\\/]+|\/(?:home|Users)\/[^/]+)/g;

function scrubPaths(input: string): string {
	return input.replace(HOME_PATH_RE, "~");
}

/**
 * Strip identifying / content-adjacent fields from an event before it leaves the
 * machine. The DEFAULT integrations are the real leak vector, so this is belt and
 * braces on top of disabling breadcrumbs + replay.
 */
function scrubEvent(event: SentryErrorEvent): SentryErrorEvent | null {
	// Honor the live opt-out gate: when off, drop the event entirely.
	if (!enabled) {
		return null;
	}
	// Remove request data (URLs/headers/cookies), any user identity, and the
	// machine hostname Sentry captures by default.
	event.request = undefined;
	event.user = undefined;
	event.server_name = undefined;
	// Scrub home-dir paths out of exception messages + frame paths.
	for (const ex of event.exception?.values ?? []) {
		if (ex.value) {
			ex.value = scrubPaths(ex.value);
		}
		for (const frame of ex.stacktrace?.frames ?? []) {
			if (frame.filename) {
				frame.filename = scrubPaths(frame.filename);
			}
			if (frame.abs_path) {
				frame.abs_path = scrubPaths(frame.abs_path);
			}
		}
	}
	return event;
}

// --- Init -------------------------------------------------------------------

/**
 * Initialize Sentry once. Safe to call when no DSN is set (graceful no-op). Uses
 * the random, account-unlinked install id; never calls setUser. Disables every
 * content-carrying default (breadcrumbs/replay/tracing) and respects the opt-out
 * from the very first event.
 */
export function initCrashReporting(): void {
	if (initialized || !SENTRY_DSN) {
		return;
	}
	initialized = true;
	sentryInit({
		dsn: SENTRY_DSN,
		release:
			(import.meta.env.VITE_APP_VERSION as string | undefined) ?? undefined,
		// No PII, no performance tracing, no profiling, no session replay.
		sendDefaultPii: false,
		tracesSampleRate: 0,
		// Drop ALL breadcrumbs. For a crash tier the stack trace is the value;
		// breadcrumbs are pure leak surface. @sentry/browser's default `breadcrumbs`
		// integration has dom + history on, emitting `ui.click` / `ui.input` (which
		// capture the clicked element's text / aria-label — i.e. conversation titles
		// or rendered agent content), plus `console` / `fetch` / `xhr` / `navigation`
		// (content + URLs). Returning null for every breadcrumb is the clean,
		// strictly-safer answer and honors the AC's "never include agent content".
		beforeBreadcrumb() {
			return null;
		},
		beforeSend: scrubEvent,
		initialScope: {
			// The only identifier: the random install id (NOT the account).
			tags: { install_id: safeInstallId() },
		},
	});
	// The opt-out is honored from the first event: scrubEvent (beforeSend) returns
	// null whenever `enabled` is false, so no separate client opt-out is needed
	// (@sentry/react has no top-level opt_out anyway).
}

function safeInstallId(): string {
	try {
		return getInstallId();
	} catch {
		return "unknown";
	}
}

/**
 * Flip the live gate from the Privacy settings toggle. Flips the in-memory flag and
 * mirrors it to localStorage for the next reload. beforeSend reads `enabled` on
 * every event, so toggling off stops egress immediately.
 */
export function setCrashReportingEnabled(next: boolean): void {
	enabled = next;
	writeEnabledMirror(next);
}

/**
 * Manually report a caught error (e.g. from an error boundary). A no-op unless a
 * DSN is configured AND the gate is on.
 */
export function reportError(error: unknown): void {
	if (!(enabled && SENTRY_DSN)) {
		return;
	}
	try {
		captureException(error);
	} catch {
		// Never let a crash-report failure surface to the user.
	}
}

File: apps/gateway/src/crash.rs
//! Crash reporting tier (#544, P3) — Sentry for Rust PANICS in the gateway.
//!
//! Mirrors Core's `crash.rs`. The gateway has no `PreferencesStore`, so it reads
//! its consent from env: `RYU_CRASH_REPORTS_ENABLED` (default ON, opt-out) plus a
//! DSN (`SENTRY_DSN`/`RYU_SENTRY_DSN`). Core's `gateway_spawn_env()` forwards the
//! user's desktop `crash-reports-enabled` pref + the DSN into these env vars when
//! it spawns the gateway as a sidecar, so the gateway tier follows the same single
//! consent toggle. With the flag off or no DSN this is a true no-op (never crashes
//! boot).
//!
//! Captures *panics only* — no `sentry-tracing`/`sentry-log`, since gateway log
//! lines can carry model/provider/prompt-adjacent data. `send_default_pii = false`,
//! `server_name = None`, and a `before_send` scrubs home-dir paths from the panic
//! message + every stack frame.
//!
//! Placement (Core vs Gateway): this reports *what crashed in the gateway* for the
//! user's own diagnostics; it enforces no policy. The DSN is a swappable config
//! value, never a hardcoded vendor.

/// The Sentry DSN env vars, in resolution order (Ryu mirror first).
const SENTRY_DSN_ENVS: [&str; 2] = ["RYU_SENTRY_DSN", "SENTRY_DSN"];

/// Env mirror of the `crash-reports-enabled` consent. Default ON (opt-out), matching
/// Core's pref default; only an explicit falsey value disables.
const CRASH_REPORTS_ENABLED_ENV: &str = "RYU_CRASH_REPORTS_ENABLED";

/// Parse a boolean env value (mirrors Core's `privacy::parse_bool`).
fn parse_bool(value: &str, default: bool) -> bool {
    match value.trim().to_ascii_lowercase().as_str() {
        "true" | "1" | "on" | "yes" => true,
        "false" | "0" | "off" | "no" => false,
        _ => default,
    }
}

/// Whether crash reporting is enabled per env (default ON, opt-out).
fn enabled() -> bool {
    std::env::var(CRASH_REPORTS_ENABLED_ENV)
        .map(|v| parse_bool(&v, true))
        .unwrap_or(true)
}

/// Resolve the configured DSN (Ryu mirror first). Empty/unset → no destination.
fn resolve_dsn() -> Option<String> {
    for key in SENTRY_DSN_ENVS {
        if let Ok(dsn) = std::env::var(key) {
            let trimmed = dsn.trim();
            if !trimmed.is_empty() {
                return Some(trimmed.to_string());
            }
        }
    }
    None
}

/// Replace the home-dir prefix in `s` with `~` so absolute paths never egress.
fn scrub_home(home: Option<&str>, s: &str) -> String {
    match home {
        Some(h) if !h.is_empty() && s.contains(h) => s.replace(h, "~"),
        _ => s.to_string(),
    }
}

fn home_dir_string() -> Option<String> {
    dirs::home_dir().map(|p| p.to_string_lossy().to_string())
}

/// Initialize Sentry crash reporting if consent (env) is on AND a DSN is set.
/// Returns the `ClientInitGuard` the caller MUST keep alive for the whole process.
/// `None` (a no-op) when disabled or unconfigured.
pub fn init() -> Option<sentry::ClientInitGuard> {
    if !enabled() {
        return None;
    }
    let dsn = resolve_dsn()?;
    Some(init_with_dsn(dsn))
}

fn init_with_dsn(dsn: String) -> sentry::ClientInitGuard {
    let home = home_dir_string();
    sentry::init((
        dsn,
        sentry::ClientOptions {
            release: Some(env!("CARGO_PKG_VERSION").into()),
            send_default_pii: false,
            server_name: None,
            before_send: Some(std::sync::Arc::new(move |mut event| {
                scrub_event(home.as_deref(), &mut event);
                Some(event)
            })),
            ..Default::default()
        },
    ))
}

/// Scrub home-dir prefixes out of the panic message + every stack-frame path.
fn scrub_event(home: Option<&str>, event: &mut sentry::protocol::Event<'static>) {
    for exception in &mut event.exception.values {
        if let Some(value) = &exception.value {
            exception.value = Some(scrub_home(home, value));
        }
        if let Some(stacktrace) = &mut exception.stacktrace {
            for frame in &mut stacktrace.frames {
                if let Some(abs) = &frame.abs_path {
                    frame.abs_path = Some(scrub_home(home, abs));
                }
                if let Some(file) = &frame.filename {
                    frame.filename = Some(scrub_home(home, file));
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn scrub_home_replaces_prefix() {
        assert_eq!(
            scrub_home(Some("/home/alice"), "panic at /home/alice/.ryu/x"),
            "panic at ~/.ryu/x"
        );
    }

    #[test]
    fn scrub_home_is_noop_without_match() {
        assert_eq!(scrub_home(None, "/home/alice/x"), "/home/alice/x");
        assert_eq!(scrub_home(Some(""), "/home/alice/x"), "/home/alice/x");
    }

    #[test]
    fn enabled_defaults_on_and_honors_falsey() {
        assert!(parse_bool("", true));
        assert!(!parse_bool("0", true));
        assert!(!parse_bool("off", true));
        assert!(parse_bool("1", false));
    }

    #[test]
    fn scrub_event_redacts_home_from_message_and_every_frame() {
        use sentry::protocol::{Event, Exception, Frame, Stacktrace};

        let home = "/home/alice";
        let mut event: Event<'static> = Event {
            exception: vec![Exception {
                ty: "panic".to_string(),
                value: Some("panic at /home/alice/.ryu/secret.toml".to_string()),
                stacktrace: Some(Stacktrace {
                    frames: vec![Frame {
                        abs_path: Some("/home/alice/src/main.rs".to_string()),
                        filename: Some("/home/alice/src/main.rs".to_string()),
                        ..Default::default()
                    }],
                    ..Default::default()
                }),
                ..Default::default()
            }]
            .into(),
            ..Default::default()
        };

        scrub_event(Some(home), &mut event);

        let ex = &event.exception.values[0];
        assert_eq!(ex.value.as_deref(), Some("panic at ~/.ryu/secret.toml"));
        let frame = &ex.stacktrace.as_ref().unwrap().frames[0];
        assert_eq!(frame.abs_path.as_deref(), Some("~/src/main.rs"));
        assert_eq!(frame.filename.as_deref(), Some("~/src/main.rs"));
    }

    #[test]
    fn scrub_event_is_a_noop_without_a_home_prefix() {
        use sentry::protocol::{Event, Exception};

        let mut event: Event<'static> = Event {
            exception: vec![Exception {
                ty: "panic".to_string(),
                value: Some("panic at /var/log/app.log".to_string()),
                ..Default::default()
            }]
            .into(),
            ..Default::default()
        };
        scrub_event(Some("/home/alice"), &mut event);
        assert_eq!(
            event.exception.values[0].value.as_deref(),
            Some("panic at /var/log/app.log")
        );
    }
}
